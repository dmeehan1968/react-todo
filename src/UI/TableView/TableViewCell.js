// @flow
import * as React from 'react'
import {
  Image,
  Text,
  View,
  StyleSheet,
} from 'react-native'

import {
  themedComponent,
} from '../../Theme'

// type Uri = string
//
// type ImageSource = {
//   uri: Uri,
//   height?: number,
//   width?: number,
// }
//
//
// type Props2 = {
//   avatar?: Uri | ImageSource | React.Node,
//   text: string | React.Node,
//   secondaryText?: string | React.Node,
//   disclosureIndicator?: boolean,
//   icon?: Uri | ImageSource | React.Node,
// }
//
//
//
export const TableViewCellAccessoryTypes = {
  none: 'none',
  disclosure: 'disclosure',
  detail: 'detail',
  detailDisclosure: 'detailDisclosure',
  checkmark: 'checkmark',
}

type TableViewCellAccessoryType = $Keys<typeof TableViewCellAccessoryTypes>

type Props = {
  name?: string,
  title: string,
  subTitle?: ?string,
  onPress?: () => void,
  onDisclosurePress?: () => void,
  onDetailPress?: () => void,
  accessoryType?: TableViewCellAccessoryType,
  // accessoryType?: 'none' | 'detail' | 'detailDisclosure' | 'disclosure' | 'checkmark',
  imageUrl?: string,

  style?: StyleSheet.StyleProp,
  primaryActionStyle?: StyleSheet.StyleProp,
  titleViewStyle?: StyleSheet.StyleProp,
  titleStyle?: StyleSheet.StyleProp,
  titleTextStyle?: StyleSheet.StyleProp,
  subTitleStyle?: StyleSheet.StyleProp,
  subTitleTextStyle?: StyleSheet.StyleProp,
  secondaryActionStyle?: StyleSheet.StyleProp,
  disclosureAccessoryTextStyle?: StyleSheet.StyleProp,
  detailAccessoryTextStyle?: StyleSheet.StyleProp,
  imageViewStyle?: StyleSheet.StyleProp,
  imageStyle?: StyleSheet.StyleProp,
}

// eslint-disable-next-line react/prefer-stateless-function
export default class TableViewCell extends React.Component<Props> {

  static defaultProps = {
    accessoryType: TableViewCellAccessoryTypes.none,
    onPress: () => undefined,
    onDisclosurePress: () => undefined,
    onDetailPress: () => undefined,
  }

  renderPrimaryAction = () => (
    <View
      name="primaryAction"
      onPress={this.props.onPress}
      style={this.props.primaryActionStyle}
    >
      {this.renderImage()}
      <View
        name="titleView"
        style={this.props.titleViewStyle}
      >
        {this.renderTitle()}
        {this.renderSubTitle()}
      </View>
    </View>
  )

  renderTitle = () => (
    <View
      name="title"
      style={this.props.titleStyle}
    >
      <Text
        style={this.props.titleTextStyle}
      >
        {this.props.title}
      </Text>
    </View>
  )

  renderSubTitle = () => {
    if (this.props.subTitle) {
      return (
        <View
          name="subTitle"
          style={this.props.subTitleStyle}
        >
          <Text style={this.props.subTitleTextStyle}>{this.props.subTitle}</Text>
        </View>
      )
    }
    return null
  }

  renderSecondaryAction = () => {
    if (this.props.accessoryType === TableViewCellAccessoryTypes.none) {
      return
    }

    return (
      <View
        name="secondaryAction"
        style={this.props.secondaryActionStyle}
      >
        {this.renderDetailIndicator()}
        {this.renderDisclosureIndicator()}
      </View>
    )
  }

  renderDisclosureIndicator = () => {
    if (this.props.accessoryType !== TableViewCellAccessoryTypes.disclosure
    && this.props.accessoryType !== TableViewCellAccessoryTypes.detailDisclosure) {
      return
    }

    return (
      <Text
        name="disclosureIndicator"
        style={this.props.disclosureAccessoryTextStyle}
        onPress={this.props.onDisclosurePress}
      >
        &gt;
      </Text>
    )
  }

  renderDetailIndicator = () => {
    if (this.props.accessoryType !== TableViewCellAccessoryTypes.detail
    && this.props.accessoryType !== TableViewCellAccessoryTypes.detailDisclosure) {
      return
    }

    return (
      <Text
        name="detailIndicator"
        style={this.props.detailAccessoryTextStyle}
        onPress={this.props.onDetailPress}
      >
        {'\u24D8'}
      </Text>
    )
  }

  renderImage = () => {
    if (!this.props.imageUrl) {
      return
    }

    return (
      <View
        name="imageView"
        style={this.props.imageViewStyle}
      >
        <Image source={{uri: this.props.imageUrl}} style={this.props.imageStyle} />
      </View>
    )
  }

  render = () => (
    <View
      name={this.props.name || 'TableViewCell'}
      style={this.props.style}
    >
      {this.renderPrimaryAction()}
      {this.renderSecondaryAction()}
    </View>
  )
}

export const StyledTableViewCell = themedComponent(TableViewCell)
