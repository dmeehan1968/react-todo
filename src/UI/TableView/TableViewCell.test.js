// @flow
import React from 'react'
import checkPropTypes from 'check-prop-types'
import { shallow, ShallowWrapper, type EnzymeSelector } from 'enzyme'
import toJson from 'enzyme-to-json'

import TableViewCell, { TableViewCellAccessoryTypes } from './TableViewCell'

class SUT {
  wrapper: ShallowWrapper

  constructor(props: {}) {
    this.wrapper = shallow(<TableViewCell title="My Title" {...props} />)
  }

  find(selector: EnzymeSelector): ShallowWrapper {
    return this.wrapper.find(selector)
  }

  toJSON(): string {
    return toJson(this.wrapper)
  }
}

describe('TableViewCell', () => {

  describe('title', () => {

    it('is required', () => {

      const result = checkPropTypes(TableViewCell.propTypes, {}, 'prop', TableViewCell.name)
      expect(result).toContain('The prop `title` is marked as required in `TableViewCell`, but its value is `undefined`')

    })

    it('renders', () => {
      const sut = new SUT({ title: 'My Title' })
      expect(sut.toJSON()).toMatchSnapshot()
    })
  })

  describe('subtitle', () => {
    it('renders', () => {
      const sut = new SUT({ subtitle: 'My Subtitle' })
      expect(sut.toJSON()).toMatchSnapshot()
    })
  })

  describe('primaryAction', () => {
    it('press', () => {
      const onPress = jest.fn()
      const sut = new SUT({ onPress })
      sut.find('[name="primaryAction"]').simulate('press')
      expect(onPress).toHaveBeenCalledTimes(1)
    })
  })

  describe('accessoryType', () => {
    it('supports none', () => {
      const sut = new SUT({ accessoryType: TableViewCellAccessoryTypes.none })
      expect(sut.toJSON()).toMatchSnapshot()
    })

    describe('disclosure', () => {

      it('press', () => {
        const onPress = jest.fn()
        const sut = new SUT({
          accessoryType: TableViewCellAccessoryTypes.disclosure,
          onDisclosurePress: onPress,
        })
        sut.find('[name="disclosureIndicator"]').simulate('press')
        expect(onPress).toHaveBeenCalledTimes(1)
      })

    })

    describe('detail', () => {
      it('press', () => {
        const onPress = jest.fn()
        const sut = new SUT({
          accessoryType: TableViewCellAccessoryTypes.detail,
          onDetailPress: onPress,
        })
        sut.find('[name="detailIndicator"]').simulate('press')
        expect(onPress).toHaveBeenCalledTimes(1)
      })
    })

    describe('disclosureDetail', () => {

      let onDisclosurePress
      let onDetailPress
      let sut

      beforeEach(() => {
        onDisclosurePress = jest.fn()
        onDetailPress = jest.fn()
        sut = new SUT({
          accessoryType: TableViewCellAccessoryTypes.detailDisclosure,
          onDisclosurePress,
          onDetailPress,
        })
      })

      it('disclosure press', () => {
        sut.find('[name="disclosureIndicator"]').simulate('press')
        expect(onDisclosurePress).toHaveBeenCalledTimes(1)
        expect(onDetailPress).toHaveBeenCalledTimes(0)
      })

      it('detail press', () => {
        sut.find('[name="detailIndicator"]').simulate('press')
        expect(onDetailPress).toHaveBeenCalledTimes(1)
        expect(onDisclosurePress).toHaveBeenCalledTimes(0)
      })
    })

  })
})
