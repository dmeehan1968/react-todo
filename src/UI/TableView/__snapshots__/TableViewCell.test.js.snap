// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableViewCell accessoryType supports none 1`] = `
<View
  name="TableViewCell"
>
  <View
    name="primaryAction"
    onPress={[Function]}
  >
    <View
      name="titleView"
    >
      <View
        name="title"
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
        >
          My Title
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`TableViewCell subtitle renders 1`] = `
<View
  name="TableViewCell"
>
  <View
    name="primaryAction"
    onPress={[Function]}
  >
    <View
      name="titleView"
    >
      <View
        name="title"
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
        >
          My Title
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`TableViewCell title renders 1`] = `
<View
  name="TableViewCell"
>
  <View
    name="primaryAction"
    onPress={[Function]}
  >
    <View
      name="titleView"
    >
      <View
        name="title"
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
        >
          My Title
        </Text>
      </View>
    </View>
  </View>
</View>
`;
